### PR TITLE
Added asserts to web steps mandatory columns

### PR DESCRIPTION
**Pull request by xsmrcek**
Fixed issue #192

### CODE REVIEW

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

**Comment by AmyKosinova** [comment]
Don't Repeat Yourself => Utility method for soft asserting.

**Comment by xsmrcek** [comment]
> Don't Repeat Yourself => Utility method for soft asserting.

What about this solution?
If you are happy with it:
What about moving it somewhere else (idk where) so we can reuse it in more steps?

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/web/WebSteps.java [hunk 1]
```diff
@@ -350,4 +358,13 @@ private String getLastOpenedWindowHandler() {
         return handles.stream().skip(handles.size() - 1).findFirst().get();
     }
 
+    private void assertMandatoryColumns(ExamplesTable examplesTable, List<String> expectedColumns) {
+        SoftAssertions softly = new SoftAssertions();
+        expectedColumns.forEach(key -> {
+            if (examplesTable.getHeaders().stream().noneMatch(column -> column.equals(key))) {
+                softly.fail("Examples table must contain column'" + key + "'");
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/web/WebSteps.java]: Nitpick but missing space between column and key? :)

**Comment by paveljandejsek** [comment]
> What about moving it somewhere else (idk where) so we can reuse it in more steps?
Maybe ExamplesTableUtil? Or ExampleTableConstraints? Dont know but I agree that it feels useful enough to be moved somewhere 🙂

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/web/WebSteps.java (hunk 1 above)
Comment (xsmrcek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/web/WebSteps.java]: Sorry didn't see this :)

**Comment by xsmrcek** [comment]
Method moved to ExamplesTableUtil.

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java [hunk 2]
```diff
@@ -142,6 +143,24 @@ public static String getValue(ExamplesTable table, String keyColumn, String key,
...
 
+    /**
+     * Checks whether examplesTable has all mandatory columns
+     *
+     * @param examplesTable     table to be checked
+     * @param expectedColumns   list of mandatory columns
+     *
+     * @throws org.assertj.core.api.SoftAssertionError   for every column missing in table
+     */
+    public static void assertMandatoryColumns(ExamplesTable examplesTable, List<String> expectedColumns) {
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]: Can you change from List<String> to String...?

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java [hunk 3]
```diff
@@ -142,6 +144,24 @@ public static String getValue(ExamplesTable table, String keyColumn, String key,
...
+     * @param examplesTable     table to be checked
+     * @param expectedColumns   list of mandatory columns
+     *
+     * @throws org.assertj.core.api.SoftAssertionError   for every column missing in table
+     */
+    public static void assertMandatoryColumns(ExamplesTable examplesTable, String ... expectedColumns) {
+        SoftAssertions softly = new SoftAssertions();
+        Arrays.stream(expectedColumns).forEach(key -> {
+            if (examplesTable.getHeaders().stream().noneMatch(column -> column.equals(key))) {
+                softly.fail("Examples table must contains column '" + key + "'");
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]:
```suggestion
                softly.fail("Examples table must contain column '" + key + "'");
```

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

**Comment by AmyKosinova** [comment]
Can we have this functionality covered by tests?

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java [hunk 4]
```diff
@@ -142,6 +144,28 @@ public static String getValue(ExamplesTable table, String keyColumn, String key,
...
+     * @param examplesTable     table to be checked
+     * @param expectedColumns   list of mandatory columns
+     *
+     * @throws org.assertj.core.api.SoftAssertionError   for every column missing in table
+     * @throws java.lang.IllegalArgumentException        when expectedColumns is missing
+     */
+    public static void assertMandatoryColumns(ExamplesTable examplesTable, String ... expectedColumns) {
+        if (expectedColumns.length < 1){
+            throw new IllegalArgumentException("expectedColumns must not be empty");
+        }
```
Comment (AmyKosinova) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]: `Assert.notEmpty(expectedColumns, "expectedColumns must not be empty");`

Comment (paveljandejsek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]: My imagining of Dan's reaction to every new comment:

Comment (xsmrcek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]: I think this reaction is more accurate:
image
But going to fix it, looks way better Amy's way.

Comment (xsmrcek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/internal/ExamplesTableUtil.java]: done :)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/208/index.xml)

**Pull request merged**